feat: Distinguish between reusable workflow inputs and trigger workflow inputs

laurentsimon opened the pull request:
closes https://github.com/slsa-framework/slsa-github-generator/issues/1737

ianlewis commented on .github/actions/verify-token/src/index.ts:
```diff
@@ -127,29 +133,40 @@ async function run(): Promise<void> {
+
+    // Extract the inputs.
+    // See https://github.com/slsa-framework/slsa-github-generator/issues/1737.
+    const rawFinalTokenObj = await filterWorkflowInputs(
```
nit: it's not actually final

```suggestion
    const rawFilteredTokenObj = await filterWorkflowInputs(
```

ianlewis commented on .github/actions/verify-token/src/utils.ts:
```diff
@@ -22,3 +27,172 @@ export function getWorkflowPath(obj: githubObj): string {
     .split("@", 1)[0];
 }
+
+export function asMap(
+  inputs: Map<string, string | number | boolean>
+): Map<string, string | number | boolean> {
+  if (!inputs) {
+    return new Map<string, string | number | boolean>();
+  }
+  if (!(inputs instanceof Map)) {
+    return new Map<string, string | number | boolean>(Object.entries(inputs));
+  }
+  return inputs;
+}
```
This kind of confuses me. If `inputs` is defined as a `Map` wouldn't it already be a `Map`? Why do we need to do a conversion?

laurentsimon replied:
It confused me too. I thought so too, but when testing it does not work. Here's what I saw in practice:

Both `JSON.parse()` and `YAML.parse()` seem to return an object that is *not* a Map:
`const rawTokenObj: rawTokenInterface = JSON.parse(rawTokenStr);`
`const workflow: GitHubWorkflowInterface = YAML.parse(content);`

even though the definition of the `rawTokenInterface` and `GitHubWorkflowInterface` are.

I also found out that TS lets me assign `object to the Map field (hose coming from the `parse()`); I am not sure why. So I need to check whether it's a map or not. And the function above, as you say, takes as input a `Map` yet I can pass it the result of `[YAML|JSON].parse()` above which is actually not a map, and it still compiles and runs properly.

I'm not sure why this is happening or even allowed in TS... But if there's a better solution, I'd like to use it

ianlewis replied:
Lets' just document this in comments for now.

ianlewis commented on .github/actions/verify-token/Makefile:
```diff
@@ -44,3 +44,4 @@ unit-test: node_modules/.installed ## Runs all unit tests.
 lint: node_modules/.installed ## Runs eslint.
 	npm run format-check
 	npm run lint
+"yaml": "^2.2.1"
```
This is a c&p miss right?

ianlewis commented on .github/actions/verify-token/src/inputs.ts:
```diff
@@ -0,0 +1,76 @@
+): rawTokenInterface {
+  const ret = Object.create(slsaToken);
+  const workflow: GitHubWorkflowInterface = YAML.parse(content);
+  slsaToken.tool.inputs = asMap(slsaToken.tool.inputs);
```
What is the type of `slsaToken.tool.inputs` initially?

laurentsimon replied:
it's an `object`, even though we declare it as a `Map` in the structure

ianlewis replied:
hmm. I guess Map and Object are basically compatible types in TS. I'll need to look up the details.

ianlewis commented on .github/actions/verify-token/src/inputs.ts:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  // Inputs defined.
+  const wInputsMap = new Map(Object.entries(workflow.on.workflow_call.inputs));
```
Would this not use `asMap` as well?

ianlewis replied:
nit: you missed this comment I think.

laurentsimon replied:
I did miss it. I updated asMap to be a generic, because the workflow inputs are a map string -> object, whereas the reusable workflow inputs are a map string -> bool | string | int.

ianlewis commented on .github/actions/verify-token/src/inputs.ts:
```diff
@@ -0,0 +1,76 @@
+import { rawTokenInterface, GitHubWorkflowInterface } from "../src/types";
+import { fetchToolWorkflow, asMap } from "./utils";
+
+export async function filterWorkflowInputs(
```
nit: add a doc comment that explains what this function is for.

ianlewis commented on .github/actions/verify-token/src/types.ts:
```diff
@@ -65,3 +65,11 @@ export interface rawTokenInterface {
   };
 }
+
+export interface GitHubWorkflowInterface {
```
nit: add doc comment
```suggestion
// GitHubWorkflowInterface represents a GitHub Actions reusable workflow.
// See: https://docs.github.com/en/actions/using-workflows/reusing-workflows#creating-a-reusable-workflow
export interface GitHubWorkflowInterface {
```

ianlewis commented on .github/actions/verify-token/src/utils.ts:
```diff
@@ -1,4 +1,9 @@
 import { githubObj } from "./types";
+import * as core from "@actions/core";
+import * as sigstore from "sigstore";
```
nit: avoid the `sigstore.sigstore` stuff.

```suggestion
import { sigstore } from "sigstore";
```

ianlewis commented on .github/actions/verify-token/src/utils.ts:
```diff
@@ -22,3 +27,172 @@ export function getWorkflowPath(obj: githubObj): string {
+    bundle.verificationMaterial.x509CertificateChain.certificates[0].rawBytes,
+    "base64"
+  );
+  const clientCertPath = "client.cert";
```
nit: Maybe consider writing to `$RUNNER_TEMP` rather than the workspace?

laurentsimon replied:
Added this comment https://github.com/slsa-framework/slsa-github-generator/issues/1918. Since this changes tscommon, I think we should resolve that issue first.

ianlewis commented on .github/actions/verify-token/src/utils.ts:
```diff
@@ -22,3 +27,172 @@ export function getWorkflowPath(obj: githubObj): string {
+  hash: string,
+  workflowPath: string
+): Promise<string> {
+  const url = `https://raw.githubusercontent.com/${repoName}/${hash}/${workflowPath}`;
```
Can you not do this with the Octokit client that's part of the GitHub Actions toolkit?

I think it would be something like:

```typescript
import * as github from "@actions/github";

const octokit = github.getOctokit(ghToken);

let result = await octokit.repos.getContent({
  owner: ownerName,
  repo: repoName,
  path: workflowPath
})

// content will be base64 encoded
const body = Buffer.from(result.data.content, 'base64').toString()
```

laurentsimon replied:
+1, the hash seems to be supported so I'll update. Thanks

laurentsimon replied:
done, ptal

ianlewis commented on .github/actions/verify-token/package.json:
```diff
@@ -25,12 +25,15 @@
     "@actions/core": "^1.10.0",
     "@actions/github": "^5.1.1",
     "@octokit/webhooks-types": "^6.9.0",
+    "node-fetch": "^2.6.9",
```
Or even better, use octokit. See comment below.

laurentsimon replied:
would love to use another dep, ill take a look at http

laurentsimon replied:
Used oktokit instead.. which uses node-fetch under the hood.

laurentsimon commented:
@ianlewis any idea what's going on with pre-submit failures? We updated cosign or another dep recently I suppose. Not urgent to merge.

ianlewis commented:
> @ianlewis any idea what's going on with pre-submit failures? We updated cosign or another dep recently I suppose. Not urgent to merge.

Not sure. The only thing touching `go.mod` recently was the update to Go 1.20. Are the changes you made to `go.mod` on this PR necessary?

laurentsimon commented:
> > @ianlewis any idea what's going on with pre-submit failures? We updated cosign or another dep recently I suppose. Not urgent to merge.
> 
> Not sure. The only thing touching `go.mod` recently was the update to Go 1.20. Are the changes you made to `go.mod` on this PR necessary?

absolutely not. I must have tried something while in the wrong folder. Reverted.
